xmtp_mls/README.md

```diff
@@ -0,0 +1,143 @@
+# XMTP MLS
+
+## Database Schema
+
+Foreign key constraints and indexes omitted for simplicity.
+
+```sql
+-- Would be a table that only holds a single identity (your own)
+CREATE TABLE identities {
+    id TEXT PRIMARY KEY NOT NULL, -- The installation_id of the identity
+    signature_keypair BLOB NOT NULL,
+    "credential" BLOB NOT NULL,
+    created_at_ns BIGINT NOT NULL,
+    wallet_address TEXT NOT NULL,
+}
+
+-- Minimum viable group schema
+CREATE TABLE groups {
+    id BLOB PRIMARY KEY NOT NULL, -- Random ID generated by group creator
+    created_at_ns BIGINT NOT NULL, -- Based on the timestamp of the welcome message
+    membership_state INT NOT NULL, -- Enum of the GROUP_MEMBERSHIP_STATE listed below
+}
+
+-- Successfully processed messages meant to be returned to the user
+CREATE TABLE group_messages {
+    id BLOB PRIMARY KEY NOT NULL -- Derived via SHA256(CONCAT(decrypted_message_bytes, conversation_id, timestamp))
+    group_id BLOB NOT NULL,
+    decrypted_message_bytes BLOB NOT NULL, -- Message contents after decryption
+    sent_at_ns BIGINT NOT NULL, -- Based on the timestamp of the message
+    sender_installation_id TEXT NOT NULL,
+    sender_wallet_address TEXT NOT NULL, -- Could remove this if we added a table mapping installation_ids to wallet addresses
+}
+
+-- Used to keep track of the last seen message timestamp in a topic
+CREATE TABLE topic_refresh_state {
+    topic TEXT PRIMARY KEY NOT NULL,
+    last_message_timestamp BIGINT NOT NULL
+}
+
+-- This table is required to retry messages that do not send successfully due to epoch conflicts
+CREATE TABLE group_intents {
+    id SERIAL PRIMARY KEY NOT NULL,
+    kind INT NOT NULL, -- INTENT_KIND
+    group_id BLOB NOT NULL,
+    "data" BLOB NOT NULL, -- Some sort of serializable blob that can be used to re-try the message if the first attempt failed due to conflict
+    "state" INT NOT NULL, -- INTENT_STATE,
+    message_hash BLOB -- The hash of the encrypted, concrete, form of the message if it was published.
```

Comment:
You may want to store the epoch in which the message is intended to land in this table. If the message lands in a subsequent epoch, it needs to be retried.

Reply:
I was kinda hoping `openmls` would handle the logic of deciding "this is from the wrong epoch" rather than us having to do the bookkeeping separately. That way it can transparently handle things like setting `max_past_epochs` for the group to a value > 0.

Reply:
OpenMLS is not going to be able to decrypt messages that you yourself sent. So when you are seeing what order messages got put in on the topic, you will not be able to "black box" run your own messages through OpenMLS to see if they ended up decryptable

Reply:
I'd love to learn more about that. My understand was that I would be able to validate my own messages, and in the case of commits be able to apply them. Is that not the case?